### ISSUE TITLE
Fixes suggested for unmodifiable private profiles

### ISSUE THREAD

**Issue by mustju** [eljeffeg/SmartCopy#78]
SmartCopy should not suggest fixes for profiles that user has no rights to modify.

**Comment by stbodie** [comment]
Traced through the code to see exactly where this bites. Writing this up properly since the original report was a one-liner.

**Problem**

SmartCopy lets a user select fields and submit changes to a Geni profile without ever indicating, up front, whether they actually have permission to edit that profile. This shows up in two different places:

1. **The popup's own import/update form** (`buildform.js`) - the field checkboxes used when creating or updating a profile have no permission awareness at all. A user can check a full set of fields, do the work of reviewing/matching a family import, hit submit, and only then discover Geni's API rejects the update because they don't have edit rights on that profile.
2. **Consistency Check** (`content.js`, the in-page warnings on Geni itself) - this part actually does have permission data available and partially uses it: several checks (`checkCase`, `checkSpace`, `checkTitle`, `checkMaidenName`, `checkSuffix`, `checkSuffixInFirstName`) already gate their clickable "[fix X]" link behind `permissions.indexOf("update-basics") !== -1`. But the warning text itself is still shown unconditionally before that check runs - so on a profile you can't edit, you still see "contains incorrect uppercase/lowercase" flagged, just with no button and no explanation for why nothing is clickable. `checkAlias` has no permission check at all, though it only links out to Geni's own edit page rather than offering an in-extension fix.

**Expected Behavior**

The user should have some clear indication - before doing the work of selecting fields and submitting - that they don't have permission to edit a given field or profile. Not a silent absence of a button, and not a failure only discovered after submitting. Ideally the affected checkbox/field is visibly disabled (greyed out) with a tooltip or label explaining why, the same way other "can't do this right now" states already get communicated elsewhere in the extension.

**Impact**

Users do real work - reviewing an import, checking the right fields, matching family members - and only find out it was pointless after clicking submit and getting a failure. That's wasted effort every time it happens, and worse on a multi-field family import where some fields may succeed and others silently don't, leaving the user unsure what actually landed on Geni.

**Potential Solution**

The data needed for this already exists and is already being fetched - every person object Geni returns includes an `actions` array (e.g. `update`, `update-basics`, `add`, `add-photo`) reflecting exactly what the current user can do to that profile. This is already read via `genifocusdata.get("actions")` / `genifamilydata[id].get("actions")` in several places in `content.js`.

- For the popup's import/update form: check `actions` for the focus profile (and, where a family member has a resolved match to an existing Geni profile via the Action picker) and disable the corresponding checkboxes/fields up front, with a tooltip explaining why - rather than letting the user find out from a failed submission. This wouldn't apply to "Add Profile" cases, since there's no existing profile to have permission on yet.
- For Consistency Check: when a check fires but permission is absent, show an explicit "no permission to fix this" indicator instead of silently omitting the fix link with no explanation.

**Comment by stbodie** [comment]
If you want this fix, tell Claude:

---

Implement a fix for eljeffeg/SmartCopy#78 - the full problem/expected-behavior/impact/solution writeup is already in the comment above on that issue, don't re-research it, just read it.

Short version: neither the popup's import/update form (`buildform.js`) nor Consistency Check (`content.js`) tell the user up front whether they actually have permission to edit a given profile - they find out only after submitting and getting a failure, or (in Consistency Check's case) see a warning with a silently-missing fix button and no explanation why.

The data needed already exists: every Geni person object includes an `actions` array (e.g. `update`, `update-basics`, `add`, `add-photo`) reflecting exactly what the current user can do to that profile, already read via `genifocusdata.get("actions")` / `genifamilydata[id].get("actions")` in several places in `content.js`.

Two parts, do them as separate commits:

1. **`buildform.js`'s import/update form.** Before rendering a field's checkbox/input as enabled, check whether the profile it would update has `update-basics` in its `actions` (`genifocusdata` for the focus profile; for a family member, only when they have a resolved match to an existing Geni profile via the Action picker built by `buildAction()` - there's nothing to check permission on for a brand-new "Add Profile" case, so skip those). If permission is absent, render the field disabled with a tooltip/label explaining why, instead of letting the user select it and find out at submit time.

2. **`content.js`'s Consistency Check.** `checkCase`, `checkSpace`, `checkTitle`, `checkMaidenName`, `checkSuffix`, and `checkSuffixInFirstName` already gate their clickable "[fix X]" link behind `permissions.indexOf("update-basics") !== -1`, but when that check fails they show nothing at all - no link, no explanation. Change that to show an explicit "no permission to fix this" indicator in that case instead of silently omitting it. `checkAlias` has no permission check at all - it only links to Geni's own edit page rather than offering an in-extension fix, so lower priority, but consider whether it should indicate permission status too.

Report back with what you find in the actual current code before making changes if anything here doesn't match what's actually there - this description is accurate as of the research that produced it, but re-verify against current `main` rather than trusting this blindly.

---

**Comment by stbodie** [comment]
Partial fix in 2a177f6: Consistency Check (\`content.js\`) now shows an explicit "no permission to fix" indicator instead of silently omitting the fix link on \`checkSpace\`, \`checkCase\`, \`checkSuffixInFirstName\`, \`checkTitle\`, \`checkMaidenName\`, and \`checkSuffix\`.

\`checkAlias\` intentionally left alone - it only links out to Geni's own edit page rather than offering an in-extension fix, so it doesn't have the same "silent missing button" problem; lower priority per the original writeup.

Still open: the bigger half of this issue - disabling the popup's import/update form checkboxes (\`buildform.js\`) up front based on \`update-basics\` permission, so users don't do the work of reviewing/selecting fields only to have the submission fail. That's a much larger change (30+ checkbox instances across a large templated form, plus family-member permission resolution via the Action picker isn't fully worked out) and needs its own pass - not tackling it in this session.

**Comment by stbodie** [comment]
**Traced how the current lock indicator actually works, in response to a user report - relevant to this issue's original ask.**

A user noticed: when a field is locked on Geni's side, a lock icon shows up next to it, but the corresponding editable input stays fully interactive - nothing actually prevents typing into it or checking its box, even though the change can't take effect. They also pointed out the inconsistency this creates with the broader case this issue describes: if the user lacks permission to edit a profile at all (or lacks a specific permission level), the form doesn't visually reflect that either, beyond one all-or-nothing check.

**What the code actually does today:**

- **Field-level locks:** `GeniPerson.isLocked()`/`lockIcon()` (`shared.js:179-203`) read Geni's own `locked_fields` data and choose `lock.png` vs `right.png`. That icon is rendered only next to the read-only "Geni Data" companion preview - it never reaches the editable input itself. `isEnabled()`/`isChecked()` (`buildform.js:1901-1976`, the functions that actually decide whether an input is disabled/checked) take `value`, `score`, `force`, and `currentValue` - lock status is never passed in or consulted anywhere in either function. So today the lock icon is purely informational: accurate, but easy to miss, and doesn't stop you from queuing up a change that Geni will just silently reject.

- **Profile-level permission:** there is a check, but it's all-or-nothing. `popup.js:783-794` blocks the *entire* form (hides the submit button, shows a warning) only when `permissions.length === 0` - zero access whatsoever. A user with *partial* access (e.g. can add relationships/photos via `add`/`add-photo` but lacks `update-basics` for editing existing fields) sees no difference in the form at all - every field still renders fully enabled.

Notably, `content.js`'s consistency-check "quickfix" links already handle the partial-permission case correctly elsewhere in this same codebase - they check `permissions.indexOf("update-basics") !== -1` per action and fall back to a plain "no permission" label otherwise. That's the right pattern; it's just not applied to the add/update form itself.

**Proposed direction (not yet implemented):** treat "this field cannot actually be saved" as one consistent state, regardless of *why* - whether that's a Geni-side field lock or the current user's own permission level - and have it do the same two things every time: show the lock icon (already correct) *and* disable the corresponding input (currently missing for locks, and missing at the per-field level for permissions too, not just the all-zero case). This would mean:
1. `isEnabled()`/`isChecked()` (or their call sites) also consult `isLocked()` for that field, forcing `disabled`/unchecked when true.
2. Fields gated behind `update-basics` (or another specific action) get disabled per-field when the current user's `permissions` array doesn't include the needed action, not just when the array is empty.

Happy to scope and implement this if there's appetite - the pieces (`isLocked()`, the `permissions` array, and the existing content.js pattern to mirror) all already exist; this is about wiring them into `isEnabled()`/`isChecked()` consistently rather than building anything new.

**Comment by stbodie** [comment]
**Lock-based inconsistency fixed in `37a5d2c` - leaving this open, since two related pieces remain.**

Implemented: a locked field on the **focus profile** now disables consistently - both the text input *and* the checkbox itself, not just the icon. This required more than wiring `isLocked()` into `isEnabled()`/`isChecked()`: the `.checknext` click handler and `applySelectAllState()` both directly re-enable a row's input based on its checkbox's own state, bypassing whatever `isEnabled()` rendered - so the checkbox itself had to become unclickable too, and `applySelectAllState()` needed guards so "select all" can't silently undo a lock either. Full detail in the commit message. Verified via a 19-assertion synthetic harness, including a diff of every changed `isChecked()`/`isEnabled()` call site to confirm nothing else shifted.

Note for the record: `parseForm()` already excludes disabled inputs from submission, so a locked field's value was never actually at risk of being saved even before this fix - this closes the *visual* inconsistency (icon shown, field still looked editable), not a data-safety gap that existed.

**Not done, still open:**
1. **This issue's original ask - permission-based disabling.** The comment above this one proposed treating "can't actually save this" as one consistent state regardless of cause (Geni lock *or* insufficient user permission), disabling per-field based on whichever `permissions` action a field requires (mirroring `content.js`'s existing `permissions.indexOf("update-basics")` pattern). This commit only covers the lock half.
2. **Family members.** They show lock icons too (`getGeniLock()`), but unlike the focus profile, their lock status isn't known until after a Geni match is selected - a live update via `setGeniFamilyData()`, not the initial render this commit touched. Same fix, different mechanism - a distinct follow-up rather than more of the same edit.

Happy to pick up either next if there's appetite.

**Comment by stbodie** [comment]
**All follow-ups landed - closing.**

Full picture across the four commits this issue ended up covering:

1. **`37a5d2c`** - a locked field (Geni-side `locked_fields`) now disables consistently on the focus profile: both the checkbox and the input, not just a decorative icon. Required disabling the checkbox itself, not just the input - the `.checknext` click handler and `applySelectAllState()` both directly re-enable an input based on its own checkbox's state, bypassing whatever was rendered at load time.
2. **`ec9cadb`** (part A) - extended the same disabling to a *missing permission*, not just a Geni-side lock, for the focus profile: `focusFieldLocked()` now also checks `genifocusdata`'s own `actions` array for `update`/`update-basics`.
3. **`b175823`** (part B) - same permission+lock consistency extended to **family members**, via `getGeniFieldLocked()` hooked into `refreshFieldCheckState()` - their lock/permission status isn't known until a Geni match is picked, a different (later, dynamic) point than the focus profile's initial render.
4. **`f68c9b3`** (part C) - directly answering the "what about Add a new profile" question: the "Add Profile" option in a family member's action dropdown is now disabled when the focus profile lacks the `add` permission, instead of only failing after submission with "Geni permission denied."

Original ask for this issue (`content.js`'s consistency-check quickfix links not suggesting edits without permission) was already correctly implemented before any of this - confirmed via `permissions.indexOf("update-basics")` gating already in place there. Everything above was the related gap a live user report surfaced: the add/update **form** itself didn't reflect the same "can't actually save this" state the way the consistency checker already did.

One known, deliberately out-of-scope corner case, documented in part C's commit: if "Add Profile" is a family member's *only* option (no existing match found) and `add` is forbidden, the option is disabled but the row can still technically be submitted - it fails server-side with the same graceful "Geni permission denied" handling that already existed, rather than being blocked in the UI beforehand. Traced doing this fully client-side first and found it leads to an unverified downstream code path; left as a known, safe (non-crashing, non-regressing) limitation rather than guessing at a fix for something not confirmed correct.

All four commits verified via synthetic Node harnesses (19 + 8 + 12 + 9 = 48 assertions total across the four), plus a full diff review of every changed `isChecked()`/`isEnabled()` call site after the first commit to confirm no unrelated logic drift.

**Comment by stbodie** [comment]
Live-tested #230's estimator changes on a real profile with a locked spouse (no `update`/`update-basics` permission) and found this issue's own "all follow-ups landed" state had real remaining gaps - noting them here since they're squarely part of what this issue set out to fix.

**`06d8cad`** - `applyProtectedDisabledState()` (the function `refreshFieldCheckState()` calls once a family member is matched to a Geni profile) only ever *disabled* a field on discovering a lock/missing permission, never un-checked it. So a field checked at initial render (before the match, and therefore the lock, was known) stayed checked-but-disabled indefinitely - exactly the state `37a5d2c` was supposed to have eliminated, just reached via a different path (post-match discovery, not initial render). Also: `about_me`/`nicknames` never passed a `locked` argument to `refreshFieldCheckState()` at all, and photo had no lock check of any kind - added `getGeniPhotoLocked()` (photo uses its own separate `add-photo` permission) and wired it in. Same commit adds a client-side pre-check for `update` in `buildTree()` (popup.js), mirroring the existing `add`/`add-photo` guards, so a no-permission profile fails with a clear message instead of reaching Geni's API and getting "Access Denied" back.

**`505630c`** - the per-field approach kept missing things (Privacy has its own separate `refreshPrivacySelect()` code path with no lock awareness at all; location sub-fields revealed only via the "show all fields" eyeball toggle weren't covered either). Replaced chasing individual fields with one blanket sweep in `setGeniFamilyData()`: when a matched profile has no `update`/`update-basics` permission at all, every checkbox and input for that member is force-disabled and unchecked in one pass, run last so it overrides everything above it (including Select All). Same commit adds a lock icon next to the family member's `Action:` dropdown, shown only when the matched profile has no edit permission, so there's a visible answer to "why did everything just grey out" instead of only the per-field icons this issue originally covered.

Both verified via synthetic Node harnesses against the real functions (7 + 9 + 8 assertions across the two commits) and `npm test`. Live-verified on the actual locked profile that triggered this - confirmed all fields, including Privacy and eyeball-revealed rows, now grey out and stay unchecked, with the new lock icon showing why.
